File: src/android/app/src/main/java/com/openminis/app/backup/BackupFileTreeExporter.kt
package com.openminis.app.backup

import com.openminis.app.logging.AppLogger
import java.io.File

/**
 * Walks a directory tree into the package: content into the blob store,
 * structure into `files.index.jsonl`. Mirrors
 * `src/ios/Agent/Backup/BackupFileTreeExporter.swift`.
 *
 * Shared by every category that carries files — Chats' `<sid>/` directories,
 * Shared Files, and Skills' attached files — so the §3.4 size cap and the dedup
 * behaviour are identical everywhere by construction, rather than by three
 * exporters agreeing to behave the same way.
 */
class BackupFileTreeExporter(
    private val blobStore: BackupBlobStore,
    private val fileIndex: BackupFileIndexWriter,
    /**
     * Snapshot cut-off. Files modified after this instant are not part of this
     * backup's view of the data.
     *
     * **Excludes only, never includes.** mtime is not always trustworthy: a
     * file restored from another backup, synced down by a provider, or copied
     * in can carry an mtime unrelated to when this device saw it, and a clock
     * change moves every comparison at once. So the rule is asymmetric on
     * purpose — a file is dropped ONLY when its mtime is clearly after the
     * cut-off, and anything unreadable or ambiguous is INCLUDED. Wrongly
     * including a slightly-too-new file costs a little package size; wrongly
     * excluding one silently loses data from the user's backup, which is the
     * failure this feature exists to prevent.
     */
    private val snapshotAtMillis: Long = Long.MAX_VALUE,
) {

    data class Result(
        var filesIncluded: Int = 0,
        var filesSkipped: Int = 0,
        var bytesIncluded: Long = 0,
        var directories: Int = 0,
        /**
         * Files excluded because they were modified after the snapshot cut-off.
         * NOT a gap in the backup — they are outside this backup's point in
         * time, so they get no tombstone.
         */
        var filesAfterSnapshot: Int = 0,
    )

    /**
     * Export everything under [root] into the package beneath [logicalPrefix].
     *
     * A missing root is not an error: a user with no skills, or a session with
     * no files, is the normal case and must not fail the whole backup.
     */
    fun export(
        root: File,
        logicalPrefix: String,
        category: BackupCategory,
        sessionId: String? = null,
    ): Result {
        val result = Result()
        if (!root.exists() || !root.isDirectory) return result

        val base = root.canonicalFile
        for (entry in base.walkTopDown()) {
            // walkTopDown yields the root itself first; it is the prefix, not
            // an entry inside the package.
            if (entry == base) continue

            val rel = entry.relativeTo(base).path.replace(File.separatorChar, '/')
            if (rel.isEmpty()) continue
            val logicalPath = "$logicalPrefix/$rel"

            // Never package a backup artifact — see isBackupArtifact.
            if (isBackupArtifact(rel)) continue

            if (entry.isDirectory) {
                // Empty directories would otherwise vanish, since nothing
                // references them.
                if (entry.listFiles()?.isEmpty() == true) {
                    fileIndex.write(BackupFileIndexEntry.directory(logicalPath, category))
                    result.directories += 1
                }
                continue
            }
            if (!entry.isFile) continue // sockets, fifos, dangling symlinks

            // Snapshot cut-off. `> 0` guards against a filesystem reporting 0
            // for an unknown mtime, which must not read as "ancient".
            val modified = entry.lastModified()
            if (modified > 0 && modified > snapshotAtMillis) {
                result.filesAfterSnapshot += 1
                continue
            }

            when (val outcome = addFile(entry, logicalPath, sessionId, category, result)) {
                is BackupBlobStore.Outcome.Stored -> {
                    fileIndex.write(
                        BackupFileIndexEntry.file(logicalPath, outcome.size, outcome.sha256, category)
                    )
                    result.filesIncluded += 1
                    result.bytesIncluded += outcome.size
                }
                is BackupBlobStore.Outcome.Duplicate -> {
                    // The bytes are already in the package under this digest;
                    // only the tree needs a second entry pointing at them.
                    fileIndex.write(
                        BackupFileIndexEntry.file(logicalPath, outcome.size, outcome.sha256, category)
                    )
                    result.filesIncluded += 1
                }
                is BackupBlobStore.Outcome.SkippedTooLarge -> {
                    // §3.4: tombstone, never a silent drop.
                    fileIndex.write(
                        BackupFileIndexEntry.sizeSkipped(logicalPath, outcome.size, category)
                    )
                    result.filesSkipped += 1
                }
                null -> {
                    // Unreadable — permissions, a file deleted mid-walk, or I/O
                    // error. Recorded as a tombstone so the gap is visible on
                    // restore instead of the file simply not being there.
                    fileIndex.write(
                        BackupFileIndexEntry.unreadable(logicalPath, entry.length(), category)
                    )
                    result.filesSkipped += 1
                }
            }
        }
        return result
    }

    private fun addFile(
        file: File,
        logicalPath: String,
        sessionId: String?,
        category: BackupCategory,
        result: Result,
    ): BackupBlobStore.Outcome? = try {
        blobStore.addFile(file, logicalPath, sessionId)
    } catch (e: Exception) {
        // One bad file must not sink an entire backup — the tombstone above
        // keeps the gap visible.
        AppLogger.error(TAG, "[Backup] unreadable file $logicalPath: ${e.message}")
        null
    }

    /**
     * True for anything that is itself a backup artifact — a delivered
     * `.minisbak`, or the directory they are delivered into.
     *
     * Without this a backup sweeps up the previous backup as "user data",
     * nesting packages inside packages until the size runs away. Matched on the
     * leading path component as well as the extension, so a whole `Backups/`
     * tree is skipped in one step rather than file by file.
     */
    private fun isBackupArtifact(rel: String): Boolean {
        if (rel.endsWith(".${BackupFormat.FILE_EXTENSION}")) return true
        return rel.substringBefore('/') == BACKUPS_DIR_NAME
    }

    companion object {
        private const val TAG = "Backup"

        /** Sibling of `shared`/`skills`/`memory`, matching iOS's delivery directory. */
        const val BACKUPS_DIR_NAME = "Backups"
    }
}

/**
 * Append-only writer for `files.index.jsonl`.
 *
 * Flushed after every record, deliberately. This file is what a resumed run
 * reads back to discover which blobs an interrupted attempt already stored
 * ([BackupBlobStore.rehydrateFromFileIndex]) — and the interruption it has to
 * survive is the process dying, which is precisely when buffered-but-unwritten
 * lines are lost. A buffer here would make resume silently forget the tail of
 * the previous attempt and re-copy blobs that are already staged.
 *
 * The cost is one small write per file rather than per 8KB; next to hashing and
 * copying the file's contents, that is noise.
 */
class BackupFileIndexWriter(file: File) : AutoCloseable {
    private val stream = file.also { it.parentFile?.mkdirs() }.outputStream()

    var written = 0
        private set

    fun write(entry: BackupFileIndexEntry) {
        val line = BackupFormat.json.encodeToString(BackupFileIndexEntry.serializer(), entry) + "\n"
        stream.write(line.toByteArray(Charsets.UTF_8))
        stream.flush()
        written += 1
    }

    override fun close() {
        stream.flush()
        stream.close()
    }
}

File: src/ios/Agent/Backup/BackupFileTreeExporter.swift
import Foundation

private let logger = AppLogger(category: "Backup")

/// Walks a directory tree into the package: content into the blob store,
/// structure into `files.index.jsonl`.
///
/// Shared by every category that carries files — Chats' `<sid>/` directories,
/// Shared Files, and Skills' attached files — so the §3.4 size cap and the
/// dedup behaviour are identical everywhere by construction rather than by
/// three exporters agreeing to behave the same way.
struct BackupFileTreeExporter {
    let blobStore: BackupBlobStore
    let fileIndex: BackupFileIndexWriter

    /// Snapshot cut-off (review S7 / snapshot task). Files modified after this
    /// instant are not part of this backup's view of the data.
    ///
    /// **Deliberately excludes only, never includes.** The task flagged that
    /// mtime is not always trustworthy, and it is right: a file restored from
    /// another backup, synced down by a FileProvider, or copied with
    /// `copyItem` can carry an mtime unrelated to when this device saw it, and
    /// a clock change moves every comparison at once. So the rule here is
    /// asymmetric on purpose — a file is dropped ONLY when its mtime is
    /// clearly after the cut-off, and anything unreadable, missing, or
    /// ambiguous is INCLUDED. Wrongly including a slightly-too-new file costs
    /// a little package size; wrongly excluding one silently loses user data
    /// from their backup, which is the failure this whole feature exists to
    /// prevent.
    var snapshotAt: Date = .distantFuture

    struct Result {
        var filesIncluded = 0
        var filesSkipped = 0
        var bytesIncluded: Int64 = 0
        var directories = 0
        /// [review S9] Subset of `filesSkipped` that was excluded because it
        /// lives in iCloud and isn't on this device. Tracked separately because
        /// the user-facing message differs from the §3.4 size cap: "too big to
        /// include" is a setting they chose, "not downloaded" is fixable by
        /// downloading the files and running the backup again.
        var filesNotDownloaded = 0
        /// Files excluded because they were modified after the snapshot
        /// cut-off. NOT a gap in the backup — they are simply outside this
        /// backup's point in time, so they get no tombstone.
        var filesAfterSnapshot = 0
    }

    /// True for anything that is itself a backup artifact — a delivered
    /// `.minisbak` package, or the directory they are delivered into.
    ///
    /// Matched on the leading path component as well as the extension so a
    /// whole `Backups/` tree is skipped in one step rather than file by file.
    private func isBackupArtifact(relativePath rel: String) -> Bool {
        if rel.hasSuffix("." + BackupFormat.fileExtension) { return true }
        let first = rel.split(separator: "/").first.map(String.init)
        return first == BackupDelivery.backupsDirectoryName
    }

    /// [review S9] Outcome of trying to materialise an iCloud/FileProvider
    /// placeholder before packaging it.
    enum DownloadOutcome {
        /// Local file, or already downloaded — package it normally.
        case ready
        /// Still a placeholder. Its bytes are NOT on this device, so packaging
        /// it would store an empty file.
        case notDownloaded(size: Int64)
    }

    /// Ensure a file's real bytes are present locally, or report that they are
    /// not.
    ///
    /// Strategy (the task left the choice open): **try to download, then
    /// degrade to a tombstone.** Requesting the download is what makes an
    /// unattended export of a mostly-offloaded library actually complete over
    /// time, but waiting on iCloud is unbounded in the worst case and a backup
    /// must not hang forever on one file — so the wait is capped and anything
    /// still absent becomes an explicit tombstone rather than a fake empty file.
    ///
    /// `MountedFolderCoordinator.ensureDownloaded` is reused deliberately: it
    /// already encapsulates the start-download-then-poll dance and its own
    /// timeout, and duplicating that logic here would let the two drift.
    /// It polls with `Thread.sleep`, which is why this must stay off the main
    /// thread — `BackupExporter` is an actor, so it already is.
    private func ensureLocallyAvailable(_ url: URL) -> DownloadOutcome {
        let keys: Set<URLResourceKey> = [.isUbiquitousItemKey,
                                         .ubiquitousItemDownloadingStatusKey,
                                         .fileSizeKey]
        guard let values = try? url.resourceValues(forKeys: keys),
              values.isUbiquitousItem == true else {
            // Not a ubiquitous item at all — the overwhelmingly common case,
            // and it costs one resourceValues call.
            return .ready
        }
        if values.ubiquitousItemDownloadingStatus == .current { return .ready }

        // A placeholder's `.fileSizeKey` is the LOGICAL size the provider
        // advertises, not 0 — that's what makes the tombstone informative.
        let logicalSize = values.fileSize.map(Int64.init) ?? 0

        do {
            try MountedFolderCoordinator.ensureDownloaded(url)
        } catch {
            logger.warning("[Backup] iCloud download failed/timed out for \(url.lastPathComponent): \(error.localizedDescription)")
            return .notDownloaded(size: logicalSize)
        }

        // ensureDownloaded returns without waiting when called on the main
        // thread, and a download can also complete partially, so re-read the
        // status rather than trusting a non-throwing return.
        let after = try? url.resourceValues(forKeys: keys)
        if after?.ubiquitousItemDownloadingStatus == .current { return .ready }
        return .notDownloaded(size: after?.fileSize.map(Int64.init) ?? logicalSize)
    }

    /// Export everything under `root`.
    ///
    /// - Parameters:
    ///   - root: host directory to walk. Missing is not an error — a session
    ///     with no attachments simply has no directory.
    ///   - logicalPrefix: package-relative prefix, e.g. `chats/<sid>`.
    ///   - category: recorded on every index line so a selective restore can
    ///     filter without re-deriving ownership from the path.
    @discardableResult
    func export(root: URL,
                logicalPrefix: String,
                category: BackupCategory,
                sessionId: String? = nil) throws -> Result {
        var result = Result()
        let fm = FileManager.default

        var isDir: ObjCBool = false
        guard fm.fileExists(atPath: root.path, isDirectory: &isDir), isDir.boolValue else {
            return result
        }

        // `.skipsHiddenFiles` matches the app's own session-tree scan
        // (ChatStore.scanAndMarkSessionFilesForResurrect) — dotfiles in these
        // directories are OS bookkeeping (.DS_Store, .Trash), not user content.
        guard let enumerator = fm.enumerator(
            at: root,
            includingPropertiesForKeys: [.isRegularFileKey, .isDirectoryKey, .fileSizeKey],
            options: [.skipsHiddenFiles]
        ) else { return result }

        for case let url as URL in enumerator {
            let values = try? url.resourceValues(forKeys: [.isRegularFileKey, .isDirectoryKey])
            // Relative path via path-prefix strip: the enumerator yields
            // absolute URLs, and `url.relativePath` is not meaningful here.
            let rel = url.path.hasPrefix(root.path + "/")
                ? String(url.path.dropFirst(root.path.count + 1))
                : url.lastPathComponent
            let logicalPath = "\(logicalPrefix)/\(rel)"

            // Never let a backup package become part of the next backup.
            //
            // Delivered packages now live OUTSIDE `shared/` (see
            // BackupDelivery.backupsDirectory), so this cannot trigger today —
            // it is deliberate defence in depth. An earlier build delivered
            // into `shared/Backups/`, which meant the next Shared Files export
            // would scan the previous package in as user data and nest packages
            // inside packages without bound. Keeping the guard means a future
            // change that puts something back under a scanned root cannot
            // silently reintroduce that.
            if isBackupArtifact(relativePath: rel) {
                if values?.isDirectory == true { enumerator.skipDescendants() }
                logger.info("[Backup] skipping backup artifact inside \(category.rawValue): \(rel)")
                continue
            }

            if values?.isDirectory == true {
                // Recorded so empty directories survive the round trip —
                // nothing else in the package references them.
                fileIndex.write(.directory(path: logicalPath, category: category))
                result.directories += 1
                continue
            }
            guard values?.isRegularFile == true else { continue }

            // Snapshot cut-off. Only a file whose mtime is CLEARLY after the
            // cut-off is dropped; see `snapshotAt` for why this is asymmetric.
            if snapshotAt != .distantFuture,
               let mtime = try? url.resourceValues(forKeys: [.contentModificationDateKey])
                   .contentModificationDate,
               mtime > snapshotAt {
                result.filesAfterSnapshot += 1
                continue
            }

            // [review S9] Before hashing anything, make sure the bytes are
            // actually on this device. An undownloaded iCloud placeholder is a
            // readable, regular, zero-byte file — indistinguishable from a real
            // empty file to every check that used to run here.
            if case .notDownloaded(let logicalSize) = ensureLocallyAvailable(url) {
                logger.warning("[Backup] skip (not downloaded from iCloud) \(logicalPath) logicalSize=\(logicalSize)")
                fileIndex.write(.notDownloaded(path: logicalPath, size: logicalSize,
                                               category: category))
                result.filesSkipped += 1
                result.filesNotDownloaded += 1
                continue
            }

            // [T-backup-scan-jetsam] Brake between files when the system is
            // under memory pressure (the 02:01 kill kept scanning straight
            // through CRITICAL), and drain autoreleased allocations per file —
            // hashing + copying is where the bytes flow, and this loop visits
            // tens of thousands of blobs in one run.
            BackupMemoryGovernor.shared.throttleIfNeeded(context: "fileTree \(category.rawValue)")
            autoreleasepool { () -> Void in
            do {
                let outcome = try blobStore.addFile(at: url, logicalPath: logicalPath,
                                                    sessionId: sessionId)
                switch outcome {
                case .stored(let sha, let size), .duplicate(let sha, let size):
                    fileIndex.write(.file(path: logicalPath, size: size,
                                          sha256: sha, category: category))
                    result.filesIncluded += 1
                    result.bytesIncluded += size
                case .skippedTooLarge(let size):
                    // §3.4: tombstone, never a silent drop. The restore side
                    // renders this as an explicit "excluded, was N bytes"
                    // placeholder instead of the file just being absent.
                    fileIndex.write(.sizeSkipped(path: logicalPath, size: size,
                                                 category: category))
                    result.filesSkipped += 1
                }
            } catch {
                // One unreadable file must not abort a multi-GB backup. Record
                // it as a tombstone so the gap is still visible, and continue.
                logger.warning("[Backup] unreadable file \(logicalPath): \(error.localizedDescription)")
                let size = (try? url.resourceValues(forKeys: [.fileSizeKey]).fileSize).flatMap { Int64($0) } ?? 0
                fileIndex.write(BackupFileIndexEntry(
                    path: logicalPath, size: size, sha256: nil,
                    category: category.rawValue, skipped: "unreadable", isDirectory: nil))
                result.filesSkipped += 1
            }
            }
        }
        return result
    }
}

/// Buffered writer for `files.index.jsonl`.
///
/// Separate from `BackupJSONLWriter` because these lines are bare entries, not
/// `t`/`v` envelopes — the file is a package-level index rather than a category
/// data stream, and the importer reads it before it knows which categories the
/// user picked.
final class BackupFileIndexWriter {
    private let url: URL
    private var handle: FileHandle?
    private let encoder: JSONEncoder = {
        let e = JSONEncoder()
        e.outputFormatting = [.sortedKeys, .withoutEscapingSlashes]
        return e
    }()
    private(set) var count = 0

    init(url: URL) {
        self.url = url
    }

    func write(_ entry: BackupFileIndexEntry) {
        do {
            if handle == nil {
                try FileManager.default.createDirectory(
                    at: url.deletingLastPathComponent(), withIntermediateDirectories: true)
                // Resume-safe: an index left by an interrupted attempt is
                // APPENDED to, not truncated. `createFile` on an existing path
                // would discard every entry the previous run had already
                // written, so the categories skipped as "already done" would
                // have no rows in files.index and their content would be
                // unreachable at restore — a package that looks complete and
                // silently isn't.
                if !FileManager.default.fileExists(atPath: url.path) {
                    FileManager.default.createFile(atPath: url.path, contents: nil)
                }
                handle = try FileHandle(forWritingTo: url)
                try handle?.seekToEnd()
            }
            var line = try encoder.encode(entry)
            line.append(0x0A)
            try handle?.write(contentsOf: line)
            count += 1
        } catch {
            logger.error("[Backup] files.index write failed for \(entry.path): \(error.localizedDescription)")
        }
    }

    func close() {
        try? handle?.close()
        handle = nil
    }
}
